### PR TITLE
ENH: improved plotting api for waterfall

### LINKED ISSUE #3808: Plot overhaul: waterfall
This issue is part of the [Visualisation overhaul](https://github.com/orgs/shap/projects/1) project.

The things we wish to overhaul are detailed in #3411, and the current status of these TODO items for this plot is tracked in the project fields (see sidebar).

### PR DESCRIPTION
**Pull request by Anushreebasics**
closes #3808 

## Summary
Added optional axis support to waterfall plot with backward-compatible parameter order:
[_waterfall.py:16]
Signature is now waterfall(shap_values, max_display=10, show=True, ax=None), with ax as the last argument to avoid positional-argument breakage.
Refactored waterfall rendering to use the explicit Matplotlib axes interface for drawing operations (plot, barh, arrow, errorbar, text, ticks, lines), instead of relying on implicit pyplot state: [_waterfall.py:109]
Preserved old sizing behavior when no axis is provided, while avoiding figure resizing when user supplies an axis [_waterfall.py:112]
Ensured show=False returns the main plotting axis (not whichever axis is currently active after twiny calls): [_waterfall.py:384]
Updated docstring to document ax and return behavior: [_waterfall.py:33]

### CODE REVIEW

**Comment by Anushreebasics** [comment]
@connortann @CloseChoice please review

**Comment by Anushreebasics** [comment]
@CloseChoice please review

File: tests/plots/test_waterfall.py [hunk 1]
```diff
@@ -118,3 +118,12 @@ def f(x):
     explainer = shap.Explainer(f, med)
     shap_values = explainer(X)
     shap.plots.waterfall(shap_values[0], show=False)
+
+
+def test_waterfall_accepts_ax_argument(explainer):
```
Comment (CloseChoice) [inline tests/plots/test_waterfall.py:124]: could you pls add `@pytest.mark.mpl_image_compare` here and return the axis so that an image gets created which we'll compare across commits?

**Review by CloseChoice — COMMENTED** [review]
thanks for the PR, this looks good. Pls address the comment below.

File: tests/plots/test_waterfall.py [hunk 2]
```diff
@@ -118,3 +118,12 @@ def f(x):
...
+    explanation = explainer(explainer.data)
+    fig, ax = plt.subplots()
+    returned_ax = shap.plots.waterfall(explanation[0], show=False, ax=ax)
+
```
Comment (CloseChoice) [inline tests/plots/test_waterfall.py:128]: could you pls also add tests for `show=True` and `ax=None`? and combinations of those?

File: shap/plots/_waterfall.py [hunk 3]
```diff
@@ -370,7 +384,7 @@ def waterfall(shap_values, max_display=10, show=True):
     if show:
         plt.show()
     else:
-        return plt.gca()
+        return ax
```
Comment (CloseChoice) [inline shap/plots/_waterfall.py:391]: hmm, isn't this different since ax3 was returned before but we are now returning just ax? Could you pls add a couple more tests verifying that everything is correct here?

File: shap/plots/_waterfall.py [hunk 4]
```diff
@@ -33,10 +33,20 @@ def waterfall(shap_values, max_display=10, show=True):
...
+    ax : matplotlib Axes, optional
+        Optionally specify an existing :external+mpl:class:`~matplotlib.axes.Axes` object, into which the
+        plot will be placed.
+
     show : bool
         Whether :external+mpl:func:`matplotlib.pyplot.show()` is called before returning.
         Setting this to ``False`` allows the plot to be customized further after it
-        has been created, returning the current axis via plt.gca().
+        has been created.
+
```
[the remark this answers is not in the document]
Comment (CloseChoice) [reply shap/plots/_waterfall.py]: pls address this as well!

**Comment by Anushreebasics** [comment]
@CloseChoice please review the changes

**Comment by Anushreebasics** [comment]
@CloseChoice  please review the changes

**Comment by Anushreebasics** [comment]
@CloseChoice @amishhaa please review

**Comment by Anushreebasics** [comment]
@CloseChoice please review
